cognite_toolkit/_cdf_tk/utils/fileio/_writers.py

```diff
@@ -0,0 +1,119 @@
+
+    def write_chunks(self, chunks: Iterable[Chunk], filestem: str = "") -> None:
+        filepath = self._get_filepath(filestem)
+        writer = self._get_writer(filepath, filestem)
+        self._write(writer, chunks)
+
+    def _get_filepath(self, filename: str) -> Path:
+        sanitized_name = f"{to_directory_compatible(filename)}-" if filename else ""
+        file_count = self._file_count_by_filename[filename]
+        file_path = (
+            self.output_dir
+            / f"{sanitized_name}part-{file_count:04}.{self.kind}{self.format}{self.compression_cls.file_suffix}"
+        )
+        file_path.parent.mkdir(parents=True, exist_ok=True)
+        return file_path
+
+    def _get_writer(self, filepath: Path, filename_base: str) -> T_IO:
+        if filepath not in self._writer_by_filepath:
+            self._writer_by_filepath[filepath] = self._create_writer(filepath)
+        elif self._is_above_file_size_limit(filepath, self._writer_by_filepath[filepath]):
+            self._writer_by_filepath[filepath].close()
+            del self._writer_by_filepath[filepath]
+            self._file_count_by_filename[filename_base] += 1
+            new_filepath = self._get_filepath(filename_base)
+            return self._get_writer(new_filepath, filename_base)
+        return self._writer_by_filepath[filepath]
+
+    def __enter__(self) -> Self:
+        self._file_count_by_filename.clear()
+        self._writer_by_filepath.clear()
+        return self
+
+    def __exit__(
+        self, exc_type: type[BaseException] | None, exc_val: BaseException | None, exc_tb: TracebackType | None
+    ) -> None:
+        for writer in self._writer_by_filepath.values():
+            writer.close()
+        self._writer_by_filepath.clear()
+        self._file_count_by_filename.clear()
+        return None
+
+    def _is_above_file_size_limit(self, filepath: Path, writer: T_IO) -> bool:
+        """Check if the file size is above the limit."""
+        try:
+            writer.flush()
+        except (AttributeError, ValueError):
+            # Some writers might not support flush (e.g. already closed).
+            # We can ignore this and proceed to check the file size on disk.
+            pass
+        return filepath.exists() and filepath.stat().st_size > self.max_file_size_bytes
+
+    @abstractmethod
+    def _create_writer(self, filepath: Path) -> T_IO:
+        """Create a writer for the given file path."""
+        raise NotImplementedError("This method should be implemented in subclasses.")
+
+    @abstractmethod
+    def _write(self, writer: T_IO, chunks: Iterable[Chunk]) -> None:
+        """Write the chunk to the file."""
+        raise NotImplementedError("This method should be implemented in subclasses.")
+
+    @classmethod
+    def create_from_format(
+        cls,
+        format: str,
+        output_dir: Path,
+        kind: str,
+        compression: type[Compression],
+    ) -> "FileWriter":
```

Comment:
Nit: I tried, and failed, to find a different way to indicate this type (since Self can't be parameterized and the parameterization is runtime, it all gets complex).  The way you did it is a way, but shouldn't the type annotation here be quoted in single quotes, and not double quotes? It is claimed that PEP-484 specifies that, but it doesn't really spell it out.

Reply:
As far as I understand single and double quotes can be used interchangeably in Python.